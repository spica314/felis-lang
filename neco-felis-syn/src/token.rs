use crate::{FileId, Pos};

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct TokenKeyword {
    pos: Pos,
    s: String,
}

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct TokenVariable {
    pos: Pos,
    s: String,
}

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct TokenOperator {
    pos: Pos,
    s: String,
}

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct TokenNumber {
    pos: Pos,
    s: String,
}

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct TokenParenL {
    pos: Pos,
}

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct TokenParenR {
    pos: Pos,
}

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct TokenBracketL {
    pos: Pos,
}

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct TokenBracketR {
    pos: Pos,
}

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct TokenBraceL {
    pos: Pos,
}

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct TokenBraceR {
    pos: Pos,
}

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct TokenComma {
    pos: Pos,
}

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct TokenColon {
    pos: Pos,
}

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct TokenSemicolon {
    pos: Pos,
}

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum Token {
    Keyword(TokenKeyword),
    Variable(TokenVariable),
    Operator(TokenOperator),
    Number(TokenNumber),
    ParenL(TokenParenL),
    ParenR(TokenParenR),
    BracketL(TokenBracketL),
    BracketR(TokenBracketR),
    BraceL(TokenBraceL),
    BraceR(TokenBraceR),
    Comma(TokenComma),
    Colon(TokenColon),
    Semicolon(TokenSemicolon),
}

fn is_operator_char(c: char) -> bool {
    ['+', '-', '*', '/', '%', '=', '<', '>'].contains(&c)
}

impl Token {
    pub fn lex(s: &str, file_id: FileId) -> Vec<Token> {
        let cs: Vec<_> = s.chars().collect();

        let mut tokens = vec![];
        let mut i = 0;
        let mut line = 1;
        let mut column = 1;

        while i < cs.len() {
            if cs[i] == '\n' {
                i += 1;
                line += 1;
                column = 1;
                continue;
            }

            while cs[i].is_whitespace() {
                i += 1;
                column += 1;
                continue;
            }

            // TokenParenL
            if cs[i] == '(' {
                let token_line = line;
                let token_column = column;
                i += 1;
                column += 1;

                let token = Token::ParenL(TokenParenL {
                    pos: Pos::new(file_id, token_line, token_column),
                });

                tokens.push(token);
                continue;
            }

            // TokenParenR
            if cs[i] == ')' {
                let token_line = line;
                let token_column = column;
                i += 1;
                column += 1;

                let token = Token::ParenR(TokenParenR {
                    pos: Pos::new(file_id, token_line, token_column),
                });

                tokens.push(token);
                continue;
            }

            // TokenBracketL
            if cs[i] == '[' {
                let token_line = line;
                let token_column = column;
                i += 1;
                column += 1;

                let token = Token::BracketL(TokenBracketL {
                    pos: Pos::new(file_id, token_line, token_column),
                });

                tokens.push(token);
                continue;
            }

            // TokenBracketR
            if cs[i] == ']' {
                let token_line = line;
                let token_column = column;
                i += 1;
                column += 1;

                let token = Token::BracketR(TokenBracketR {
                    pos: Pos::new(file_id, token_line, token_column),
                });

                tokens.push(token);
                continue;
            }

            // TokenBraceL
            if cs[i] == '{' {
                let token_line = line;
                let token_column = column;
                i += 1;
                column += 1;

                let token = Token::BraceL(TokenBraceL {
                    pos: Pos::new(file_id, token_line, token_column),
                });

                tokens.push(token);
                continue;
            }

            // TokenBraceR
            if cs[i] == '}' {
                let token_line = line;
                let token_column = column;
                i += 1;
                column += 1;

                let token = Token::BraceR(TokenBraceR {
                    pos: Pos::new(file_id, token_line, token_column),
                });

                tokens.push(token);
                continue;
            }

            // TokenComma
            if cs[i] == ',' {
                let token_line = line;
                let token_column = column;
                i += 1;
                column += 1;

                let token = Token::Comma(TokenComma {
                    pos: Pos::new(file_id, token_line, token_column),
                });

                tokens.push(token);
                continue;
            }

            // TokenColon
            if cs[i] == ':' {
                let token_line = line;
                let token_column = column;
                i += 1;
                column += 1;

                let token = Token::Colon(TokenColon {
                    pos: Pos::new(file_id, token_line, token_column),
                });

                tokens.push(token);
                continue;
            }

            // TokenSemicolon
            if cs[i] == ';' {
                let token_line = line;
                let token_column = column;
                i += 1;
                column += 1;

                let token = Token::Semicolon(TokenSemicolon {
                    pos: Pos::new(file_id, token_line, token_column),
                });

                tokens.push(token);
                continue;
            }

            // TokenKeyword
            if cs[i] == '#' {
                let token_line = line;
                let token_column = column;
                i += 1;
                column += 1;
                let mut buf = String::new();
                while i < cs.len() && cs[i].is_ascii_alphanumeric() {
                    buf.push(cs[i]);
                    i += 1;
                    column += 1;
                }

                let token = Token::Keyword(TokenKeyword {
                    s: buf,
                    pos: Pos::new(file_id, token_line, token_column),
                });

                tokens.push(token);
                continue;
            }

            // TokenNumber
            if cs[i].is_numeric() {
                let token_line = line;
                let token_column = column;
                let mut buf = String::new();
                while i < cs.len()
                    && (cs[i].is_ascii_alphanumeric() || cs[i] == '.' || cs[i] == '_')
                {
                    buf.push(cs[i]);
                    i += 1;
                    column += 1;
                }

                let token = Token::Number(TokenNumber {
                    pos: Pos::new(file_id, token_line, token_column),
                    s: buf,
                });

                tokens.push(token);
                continue;
            }

            // TokenVariable
            if cs[i].is_ascii_alphabetic() || cs[i] == '_' {
                let token_line = line;
                let token_column = column;
                let mut buf = String::new();
                while i < cs.len()
                    && (cs[i].is_ascii_alphanumeric() || cs[i] == '.' || cs[i] == '_')
                {
                    buf.push(cs[i]);
                    i += 1;
                    column += 1;
                }

                let token = Token::Variable(TokenVariable {
                    pos: Pos::new(file_id, token_line, token_column),
                    s: buf,
                });

                tokens.push(token);
                continue;
            }

            // TokenVariable
            if is_operator_char(cs[i]) {
                let token_line = line;
                let token_column = column;
                let mut buf = String::new();
                while i < cs.len() && is_operator_char(cs[i]) {
                    buf.push(cs[i]);
                    i += 1;
                    column += 1;
                }

                let token = Token::Operator(TokenOperator {
                    pos: Pos::new(file_id, token_line, token_column),
                    s: buf,
                });

                tokens.push(token);
                continue;
            }

            panic!("line = {line}, column = {column}");
        }

        tokens
    }
}

#[cfg(test)]
mod test {
    use crate::FileIdGenerator;

    use super::*;
    use insta::assert_debug_snapshot;

    #[test]
    fn test_lex_eq_and_nat() {
        let mut file_id_generator = FileIdGenerator::new();
        let file_id = file_id_generator.generate_file_id();
        let s = std::fs::read_to_string("../testcases/felis/single/eq_and_nat.fe").unwrap();
        let tokens = Token::lex(&s, file_id);

        assert_debug_snapshot!(tokens);
    }
}
