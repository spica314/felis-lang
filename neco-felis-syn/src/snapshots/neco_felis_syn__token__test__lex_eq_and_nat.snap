---
source: neco-felis-syn/src/token.rs
expression: tokens
---
[
    Keyword(
        TokenKeyword {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 1,
                column: 1,
            },
            s: "inductive",
        },
    ),
    Variable(
        TokenVariable {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 1,
                column: 12,
            },
            s: "eq",
        },
    ),
    Colon(
        TokenColon {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 1,
                column: 15,
            },
        },
    ),
    ParenL(
        TokenParenL {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 1,
                column: 17,
            },
        },
    ),
    Variable(
        TokenVariable {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 1,
                column: 18,
            },
            s: "A",
        },
    ),
    Colon(
        TokenColon {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 1,
                column: 20,
            },
        },
    ),
    Variable(
        TokenVariable {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 1,
                column: 22,
            },
            s: "Type",
        },
    ),
    ParenR(
        TokenParenR {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 1,
                column: 26,
            },
        },
    ),
    Operator(
        TokenOperator {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 1,
                column: 28,
            },
            s: "->",
        },
    ),
    ParenL(
        TokenParenL {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 1,
                column: 31,
            },
        },
    ),
    Variable(
        TokenVariable {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 1,
                column: 32,
            },
            s: "x",
        },
    ),
    Colon(
        TokenColon {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 1,
                column: 34,
            },
        },
    ),
    Variable(
        TokenVariable {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 1,
                column: 36,
            },
            s: "A",
        },
    ),
    ParenR(
        TokenParenR {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 1,
                column: 37,
            },
        },
    ),
    Operator(
        TokenOperator {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 1,
                column: 39,
            },
            s: "->",
        },
    ),
    Variable(
        TokenVariable {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 1,
                column: 42,
            },
            s: "A",
        },
    ),
    Operator(
        TokenOperator {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 1,
                column: 44,
            },
            s: "->",
        },
    ),
    Variable(
        TokenVariable {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 1,
                column: 47,
            },
            s: "Prop",
        },
    ),
    BraceL(
        TokenBraceL {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 1,
                column: 52,
            },
        },
    ),
    Variable(
        TokenVariable {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 2,
                column: 5,
            },
            s: "eq_refl",
        },
    ),
    Colon(
        TokenColon {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 2,
                column: 12,
            },
        },
    ),
    ParenL(
        TokenParenL {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 2,
                column: 14,
            },
        },
    ),
    Variable(
        TokenVariable {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 2,
                column: 15,
            },
            s: "A",
        },
    ),
    Colon(
        TokenColon {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 2,
                column: 17,
            },
        },
    ),
    Variable(
        TokenVariable {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 2,
                column: 19,
            },
            s: "Type",
        },
    ),
    ParenR(
        TokenParenR {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 2,
                column: 23,
            },
        },
    ),
    Operator(
        TokenOperator {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 2,
                column: 25,
            },
            s: "->",
        },
    ),
    ParenL(
        TokenParenL {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 2,
                column: 28,
            },
        },
    ),
    Variable(
        TokenVariable {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 2,
                column: 29,
            },
            s: "x",
        },
    ),
    Colon(
        TokenColon {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 2,
                column: 31,
            },
        },
    ),
    Variable(
        TokenVariable {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 2,
                column: 33,
            },
            s: "A",
        },
    ),
    ParenR(
        TokenParenR {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 2,
                column: 34,
            },
        },
    ),
    Operator(
        TokenOperator {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 2,
                column: 36,
            },
            s: "->",
        },
    ),
    Variable(
        TokenVariable {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 2,
                column: 39,
            },
            s: "eq",
        },
    ),
    Variable(
        TokenVariable {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 2,
                column: 42,
            },
            s: "A",
        },
    ),
    Variable(
        TokenVariable {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 2,
                column: 44,
            },
            s: "x",
        },
    ),
    Variable(
        TokenVariable {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 2,
                column: 46,
            },
            s: "x",
        },
    ),
    BraceR(
        TokenBraceR {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 3,
                column: 1,
            },
        },
    ),
    Keyword(
        TokenKeyword {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 5,
                column: 1,
            },
            s: "inductive",
        },
    ),
    Variable(
        TokenVariable {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 5,
                column: 12,
            },
            s: "nat",
        },
    ),
    Colon(
        TokenColon {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 5,
                column: 16,
            },
        },
    ),
    Variable(
        TokenVariable {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 5,
                column: 18,
            },
            s: "Set",
        },
    ),
    BraceL(
        TokenBraceL {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 5,
                column: 22,
            },
        },
    ),
    Variable(
        TokenVariable {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 6,
                column: 5,
            },
            s: "O",
        },
    ),
    Colon(
        TokenColon {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 6,
                column: 6,
            },
        },
    ),
    Variable(
        TokenVariable {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 6,
                column: 8,
            },
            s: "nat",
        },
    ),
    Comma(
        TokenComma {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 6,
                column: 11,
            },
        },
    ),
    Variable(
        TokenVariable {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 7,
                column: 5,
            },
            s: "S",
        },
    ),
    Colon(
        TokenColon {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 7,
                column: 6,
            },
        },
    ),
    Variable(
        TokenVariable {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 7,
                column: 8,
            },
            s: "nat",
        },
    ),
    Operator(
        TokenOperator {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 7,
                column: 12,
            },
            s: "->",
        },
    ),
    Variable(
        TokenVariable {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 7,
                column: 15,
            },
            s: "nat",
        },
    ),
    Comma(
        TokenComma {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 7,
                column: 18,
            },
        },
    ),
    BraceR(
        TokenBraceR {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 8,
                column: 1,
            },
        },
    ),
    Keyword(
        TokenKeyword {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 10,
                column: 1,
            },
            s: "definition",
        },
    ),
    Variable(
        TokenVariable {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 10,
                column: 13,
            },
            s: "add",
        },
    ),
    Colon(
        TokenColon {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 10,
                column: 17,
            },
        },
    ),
    ParenL(
        TokenParenL {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 10,
                column: 19,
            },
        },
    ),
    Variable(
        TokenVariable {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 10,
                column: 20,
            },
            s: "n",
        },
    ),
    Colon(
        TokenColon {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 10,
                column: 22,
            },
        },
    ),
    Variable(
        TokenVariable {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 10,
                column: 24,
            },
            s: "nat",
        },
    ),
    ParenR(
        TokenParenR {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 10,
                column: 27,
            },
        },
    ),
    Operator(
        TokenOperator {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 10,
                column: 29,
            },
            s: "->",
        },
    ),
    ParenL(
        TokenParenL {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 10,
                column: 32,
            },
        },
    ),
    Variable(
        TokenVariable {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 10,
                column: 33,
            },
            s: "m",
        },
    ),
    Colon(
        TokenColon {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 10,
                column: 35,
            },
        },
    ),
    Variable(
        TokenVariable {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 10,
                column: 37,
            },
            s: "nat",
        },
    ),
    ParenR(
        TokenParenR {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 10,
                column: 40,
            },
        },
    ),
    Operator(
        TokenOperator {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 10,
                column: 42,
            },
            s: "->",
        },
    ),
    Variable(
        TokenVariable {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 10,
                column: 45,
            },
            s: "nat",
        },
    ),
    BraceL(
        TokenBraceL {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 10,
                column: 49,
            },
        },
    ),
    Keyword(
        TokenKeyword {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 11,
                column: 5,
            },
            s: "match",
        },
    ),
    Variable(
        TokenVariable {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 11,
                column: 12,
            },
            s: "n",
        },
    ),
    BraceL(
        TokenBraceL {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 11,
                column: 14,
            },
        },
    ),
    Variable(
        TokenVariable {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 12,
                column: 9,
            },
            s: "O",
        },
    ),
    Operator(
        TokenOperator {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 12,
                column: 11,
            },
            s: "=>",
        },
    ),
    Variable(
        TokenVariable {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 12,
                column: 14,
            },
            s: "m",
        },
    ),
    Comma(
        TokenComma {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 12,
                column: 15,
            },
        },
    ),
    Variable(
        TokenVariable {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 13,
                column: 9,
            },
            s: "S",
        },
    ),
    Variable(
        TokenVariable {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 13,
                column: 11,
            },
            s: "p",
        },
    ),
    Operator(
        TokenOperator {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 13,
                column: 13,
            },
            s: "=>",
        },
    ),
    Variable(
        TokenVariable {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 13,
                column: 16,
            },
            s: "S",
        },
    ),
    ParenL(
        TokenParenL {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 13,
                column: 18,
            },
        },
    ),
    Variable(
        TokenVariable {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 13,
                column: 19,
            },
            s: "add",
        },
    ),
    Variable(
        TokenVariable {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 13,
                column: 23,
            },
            s: "p",
        },
    ),
    Variable(
        TokenVariable {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 13,
                column: 25,
            },
            s: "m",
        },
    ),
    ParenR(
        TokenParenR {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 13,
                column: 26,
            },
        },
    ),
    BraceR(
        TokenBraceR {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 14,
                column: 5,
            },
        },
    ),
    BraceR(
        TokenBraceR {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 15,
                column: 1,
            },
        },
    ),
    Keyword(
        TokenKeyword {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 17,
                column: 1,
            },
            s: "theorem",
        },
    ),
    Variable(
        TokenVariable {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 17,
                column: 10,
            },
            s: "add_0_1_eq_add_1_0",
        },
    ),
    Colon(
        TokenColon {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 17,
                column: 29,
            },
        },
    ),
    Variable(
        TokenVariable {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 17,
                column: 31,
            },
            s: "eq",
        },
    ),
    Variable(
        TokenVariable {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 17,
                column: 34,
            },
            s: "nat",
        },
    ),
    ParenL(
        TokenParenL {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 17,
                column: 38,
            },
        },
    ),
    Variable(
        TokenVariable {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 17,
                column: 39,
            },
            s: "add",
        },
    ),
    Variable(
        TokenVariable {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 17,
                column: 43,
            },
            s: "O",
        },
    ),
    ParenL(
        TokenParenL {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 17,
                column: 45,
            },
        },
    ),
    Variable(
        TokenVariable {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 17,
                column: 46,
            },
            s: "S",
        },
    ),
    Variable(
        TokenVariable {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 17,
                column: 48,
            },
            s: "O",
        },
    ),
    ParenR(
        TokenParenR {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 17,
                column: 49,
            },
        },
    ),
    ParenR(
        TokenParenR {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 17,
                column: 50,
            },
        },
    ),
    ParenL(
        TokenParenL {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 17,
                column: 52,
            },
        },
    ),
    Variable(
        TokenVariable {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 17,
                column: 53,
            },
            s: "add",
        },
    ),
    ParenL(
        TokenParenL {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 17,
                column: 57,
            },
        },
    ),
    Variable(
        TokenVariable {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 17,
                column: 58,
            },
            s: "S",
        },
    ),
    Variable(
        TokenVariable {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 17,
                column: 60,
            },
            s: "O",
        },
    ),
    ParenR(
        TokenParenR {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 17,
                column: 61,
            },
        },
    ),
    Variable(
        TokenVariable {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 17,
                column: 63,
            },
            s: "O",
        },
    ),
    ParenR(
        TokenParenR {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 17,
                column: 64,
            },
        },
    ),
    BraceL(
        TokenBraceL {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 17,
                column: 66,
            },
        },
    ),
    Variable(
        TokenVariable {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 18,
                column: 5,
            },
            s: "eq_refl",
        },
    ),
    Variable(
        TokenVariable {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 18,
                column: 13,
            },
            s: "nat",
        },
    ),
    ParenL(
        TokenParenL {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 18,
                column: 17,
            },
        },
    ),
    Variable(
        TokenVariable {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 18,
                column: 18,
            },
            s: "S",
        },
    ),
    Variable(
        TokenVariable {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 18,
                column: 20,
            },
            s: "O",
        },
    ),
    ParenR(
        TokenParenR {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 18,
                column: 21,
            },
        },
    ),
    BraceR(
        TokenBraceR {
            pos: Pos {
                file_id: FileId(
                    0,
                ),
                line: 19,
                column: 1,
            },
        },
    ),
]
